// squint/test/analyzer.js
var assert = require( 'assert' )
  , Analyzer = require( 'squint/analyzer' ).Analyzer
  , Parser = require( 'squint/parser' ).Parser 
  , events = require( 'events' )
  , makeEmitTester = require( 'squint/test/tester' ).makeEmitTester;

function runTest() {

  typeTemplateDeclaration();
 /* 
  typeDeclaration();
  templatesAsFunctionParameters();
  functionSignatures();
  functionLikeMacrosAsTemplateParameter();
  templateParameters();
*/
}

function typeTemplateDeclaration() {

  var parser = makeEmitTester( new Parser() )
    , analyzer = makeEmitTester( new Analyzer( parser ) );
  
  parser.expect( 'open', 'template<class, class> type XYZ' );
  analyzer.expect( 'template parameters', 'template<class, class>' );
  analyzer.expect( 'type declaration', 'type XYZ' );
  parser.process( 'template<class, class> type XYZ{' );
}

function typeDeclaration() {
  var parser = makeEmitTester( new Parser() )
    , analyzer = makeEmitTester( new Analyzer( parser ) );
  
  parser.expect( 'open', 'type XYZ' );
  analyzer.expect( 'type declaration', 'type XYZ' );
  parser.process( 'type XYZ{' );
}

function templatesAsFunctionParameters() {
  var parser = makeEmitTester( new Parser() )
    , analyzer = makeEmitTester( new Analyzer( parser ) );

  parser.expect( 'open', 'type foo( st< abc > )' );
  analyzer.expect( 'function signature', 'type foo( st< abc > )' );
  parser.process( 'type foo( st< abc > ){' );
}

function functionSignatures() {

  var parser = makeEmitTester( new Parser() )
    , analyzer = makeEmitTester( new Analyzer( parser ) );

  parser.expect( 'open', 'type foo()' );
  analyzer.expect( 'function signature', 'type foo()' );
  parser.process( 'type foo(){' );
}

function functionLikeMacrosAsTemplateParameter() {

  var parser = makeEmitTester( new Parser() )
    , analyzer = makeEmitTester( new Analyzer( parser ) );
    
  parser.expect( 'open', 'template class< MACRO( arg ) >' );
  analyzer.expect( 'template parameters', 'template class< MACRO( arg ) >' );
  parser.process( 'template class< MACRO( arg ) >{' );

  parser.expect( 'open', 'template class< MACRO( arg ) > class C' );
  analyzer.expect( 'template parameters', 'template class< MACRO( arg ) >' );
  parser.process( 'template class< MACRO( arg ) > class C{' );

  parser.expect( 'open', 'template class< MACRO( arg ), U > class C' );
  analyzer.expect( 'template parameters', 'template class< MACRO( arg ), U >' );
  parser.process( 'template class< MACRO( arg ), U > class C{' );
  
  parser.expect( 'open', 'template class< MACRO( arg ), template <class U> class > class C' );
  analyzer.expect( 'template parameters', 'template class< MACRO( arg ), template <class U> class >' );
  parser.process( 'template class< MACRO( arg ), template <class U> class > class C{' );
}; 

function templateParameters() {

  var parser = makeEmitTester( new Parser() )
    , analyzer = makeEmitTester( new Analyzer( parser ) );

  parser.expect( 'open', 'template class< arg >' );
  analyzer.expect( 'template parameters', 'template class< arg >' );
  parser.process( 'template class< arg >{' );
}

module.exports = {
  run : runTest
};
