// squint/test/parser.js
var assert = require( 'assert' )
  , Parser = require( 'squint/parser' ).Parser;

function TestParser( map ) {
  var expectations = [];

  Parser.call( this, map );
  
  process.on( 'exit', function() {
    assert.equal( expectations.length, 0 );
  } );

  this.expect = function( event, code ) {
    if (!expectations.length) {
      this.once( event, check );
    }
    expectations.push( { event: event, code: code } );
  }; 

  function check( code ) {
    var expectation = expectations[0];
    expectations.splice( 0, 1 );
   
    if (expectation.code != undefined) {
      assert.deepEqual( code, expectation.code );
    }
    
    if (expectations.length) {
      this.once( expectations[0].event, check );
    }
  }
};

TestParser.prototype = new Parser();

module.exports = { 
  run : function() {
    
    functionParameters();
    
    testNested();
    testInterleaved();
    testScopes();
    testStatments();
  }, 
  Parser: TestParser
};

function functionParameters() {
  var parser = new TestParser( { 
      '(': 'open parameter', ')': 'close parameter', 
      '<': 'open parameter', '>': 'close parameter' 
    } );
  
  parser.expect( 'open parameter', 'a b' );
  parser.process( 'a b(' );

  parser.expect( 'open parameter', 'a b' );
  parser.expect( 'close parameter', 'abc' );
  parser.process( 'a b( abc )' );
  
  parser.expect( 'open parameter', 'a b' );
  parser.expect( 'close parameter', 'abc' );
  parser.process( 'a b< abc >' );
}

function testNested() {
  var parser = new TestParser();

  parser.expect( 'open', 'a' );
  parser.expect( 'open', 'b' );
  parser.expect( 'statement', 'c' );
  parser.expect( 'close' );
  parser.expect( 'close' );

  parser.process( 'a { b { c; } }' );
};

function testInterleaved() { 
  var parser = new TestParser();
  
  parser.expect( 'open', 'a' );
  parser.expect( 'statement','b' );
  parser.expect( 'close' );
   
  parser.process( 'a{ b; }' );
}

function testScopes() {
  var parser = new TestParser();
  
  parser.expect( 'open', 'a' ); 
  parser.expect( 'close' );
  parser.expect( 'open', 'b' );
  parser.expect( 'close' );
  parser.expect( 'open', 'c' );
  parser.expect( 'close' );
 
  parser.process( 'a{} b{} c{}' );
}

function testStatments() {
  var parser = new TestParser();

  parser.expect( 'statement', '1' );
  parser.expect( 'statement', '2' );
  parser.expect( 'statement', '3' );
  parser.expect( 'statement' );
 
  parser.process( '1; 2; 3;;' );
}
