// squint/squint.js
/* 
note: order matters
	i.e. functionDeclare only works after typeDeclares have been removed
*/ 

var commentSingle = new RegExp( '\/\/.*\n?', 'g' )
  , commentMultiple = new RegExp( '\/\\*.*?\\*\/', 'mg' )
  , include = new RegExp( '#.*include.*\n?', 'g' )
  , defineNewLine = new RegExp( '\\\\[ \t]*\n', 'mg' )
  , define = new RegExp( '#.*define.*\s*\n?', 'g' )
  , undefine = new RegExp( '#.*undef.*\n?', 'mg' )
  , typeDeclare = new RegExp( 'class[^\{]*;|template[^\{]*;|struct[^\{]*;' )
  , functionDeclare = new RegExp( '.*?\\\)[ \t]*?;' )
  , stringLiteral = new RegExp( '".*?([^\\\\]")', 'g' )
  , arrayInitBlock = RegExp( '\\s*=.*?;', 'g' )
  , Parser = require( 'squint/parser' ).Parser
  , Builder = require( 'squint/builder' ).Builder
  , Factory = require( 'squint/factory' ).Factory
  , events = require( 'events' );
  
module.exports.stripArrayInitializerBlocks = function( code ) {
  code = code.replace( arrayInitBlock, ';' );
  return code;
};
  
module.exports.declare = function( code, context ) {

  var parser = new Parser()
    , builder = new Builder( parser, new Factory() );

  code = module.exports.stripComments( code );
  code = module.exports.stripIncludes( code );
  code = module.exports.stripDefines( code );
  code = module.exports.stripStrings( code );
  code = parser.process( code );

  return code;
};

module.exports.defineFunctions = function( code ) {
  code = module.exports.stripComments( code );
  code = module.exports.stripIncludes( code );
  code = module.exports.stripDefines( code );
  code = module.getFunctionDeclares( code );
  
  code = code.replace( ';', '{}' );
  return code;
};
    
module.exports.getFunctionDeclares = function( code ) {
  code = code.match( functionDeclare );
  if (!code) {
    code = '';
  }
  return code;
};  

module.exports.getTypeDeclares = function( code ) {
  code = code.match( typeDeclare );
  if (!code) {
    code = '';
  }
  return code;
};

module.exports.stripStrings = function( code ) { 
  code = code.replace( stringLiteral, '' );
  return code;
};

module.exports.stripDefines = function( code ) {
  code = code.replace( defineNewLine, '' );
  code = code.replace( define, '' );
  code = code.replace( undefine, '' );
  return code;
};

module.exports.stripIncludes = function( code ) {
  code = code.replace( include, '' );
  return code;
}; 

module.exports.stripComments = function( code ) {
  code = code.replace( commentSingle, '' );
  code = code.replace( commentMultiple, '' );
  return code;
};
